Menus/ObjectPickerGUI.cpp

```diff
@@ -25,1296 +12,559 @@
+			return dynamic_cast<const SceneObject *>(selectedItem->m_pEntity);
+		}
+		return nullptr;
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	void ObjectPickerGUI::SelectObjectByIndex(int objectIndex, bool playSelectionSound) {
+		m_SelectedObjectIndex = objectIndex;
+		m_ObjectsList->SetSelectedIndex(m_SelectedObjectIndex);
+		if (playSelectionSound) { g_GUISound.SelectionChangeSound()->Play(m_Controller->GetPlayer()); }
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	void ObjectPickerGUI::SelectNextOrPrevObject(bool getPrev) {
+		int objectIndex = m_SelectedObjectIndex;
+		if (getPrev) {
+			objectIndex--;
+			if (objectIndex < 0) { objectIndex = m_ObjectsList->GetItemList()->size() - 1; }
+		} else {
+			objectIndex++;
+			if (objectIndex >= m_ObjectsList->GetItemList()->size()) { objectIndex = 0; }
+		}
+		SelectObjectByIndex(objectIndex);
+	}
 
-				if(anEvent.GetControl()->GetName() == "btnMove")
-                {
-					GUIListBox *LB1 = (GUIListBox *)m_pGUIController->GetControl("list1");
-					GUIListBox *LB2 = (GUIListBox *)m_pGUIController->GetControl("list2");
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 
-					vector<GUIListPanel::Item *> *List = LB1->GetSelectionList();
-					vector<GUIListPanel::Item *>::iterator it;
+	void ObjectPickerGUI::AddObjectsListModuleGroup(int moduleID) {
+		const DataModule *dataModule = g_PresetMan.GetDataModule(moduleID);
+		std::string moduleName = dataModule->GetFriendlyName();
+		std::transform(moduleName.begin(), moduleName.end(), moduleName.begin(), ::toupper);
+		GUIBitmap *dataModuleIcon = dataModule->GetIcon() ? new AllegroBitmap(dataModule->GetIcon()) : nullptr;
+		m_ObjectsList->AddItem(moduleName, m_ExpandedModules.at(moduleID) ? "-" : "+", dataModuleIcon, nullptr, moduleID);
+	}
 
-					// Add the items in the selection list to the second list box
-					for(it = List->begin(); it != List->end(); it++)
-                    {
-						GUIListPanel::Item *I = *it;
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 
-						LB2->AddItem(I->m_Name, I->m_Tag);
+	void ObjectPickerGUI::SetObjectsListModuleGroupExpanded(int moduleID, bool expanded) {
+		if (moduleID > 0 && moduleID < g_PresetMan.GetTotalModuleCount()) {
+			m_ExpandedModules.at(moduleID) = expanded;
+			UpdateObjectsList(false);
+		} else {
+			std::fill(m_ExpandedModules.begin(), m_ExpandedModules.end(), expanded);
+		}
+	}
 
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	void ObjectPickerGUI::ToggleObjectsListModuleGroupExpansion(int moduleID) {
+		m_ExpandedModules.at(moduleID) = !m_ExpandedModules.at(moduleID);
```

Comment:
Not a big deal, but this should probably have a safety check on it to ensure `moduleID > 0 && moduleID < g_PresetMan.GetTotalModuleCount()`
I guess it doesn't really need it in its current uses, but it's still probably smart to do.